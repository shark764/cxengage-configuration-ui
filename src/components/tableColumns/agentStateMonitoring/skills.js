/*
 * Copyright Â© 2015-2019 Serenova, LLC. All rights reserved.
 */

import React from 'react';
import PropTypes from 'prop-types';
import CheckboxFilterMenu from '../../../containers/CheckboxFilterMenu';

export const helperFunctions = {
  skillsFilter: (onChange, tableType, allActive) => skillsFilter(onChange, tableType, allActive),
  filterMethod: ({ value = 'All' }, { _original: { skillIds } }, filterValues, allActive) =>
    value === 'All' || allActive
      ? allActive
      : skillIds.filter(value => filterValues.map(({ id }) => id).includes(value)).length > 0
};

export default function(skills, tableType, filterValues, allActive) {
  const skillsColumn = {
    id: 'skills',
    Header: 'Skills',
    show: skills,
    sortable: false,
    minWidth: 130,
    accessor: d => d.skills.map(skill => skill.skillName).join(', '),
    filterMethod: (filter, row) => helperFunctions.filterMethod(filter, row, filterValues, allActive),
    Filter: ({ onChange }) => helperFunctions.skillsFilter(onChange, tableType, allActive),
    Cell: ({ value }) =>
      value ? (
        <span title={value}>{value}</span>
      ) : (
        <div style={{ textAlign: 'center' }}>
          <span>--</span>
        </div>
      )
  };

  skillsColumn.Cell.propTypes = { value: PropTypes.any };

  return skillsColumn;
}

export function skillsFilter(onChange, tableType, allActive) {
  return (
    <CheckboxFilterMenu
      menuType="Skills"
      tableType={tableType}
      buttonType="columnFilter"
      selectionType="checkbox"
      updateFilter={onChange}
      hasActiveFilter={!allActive}
    >
      Skills
    </CheckboxFilterMenu>
  );
}
