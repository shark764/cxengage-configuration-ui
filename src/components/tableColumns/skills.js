/*
 * Copyright Â© 2015-2018 Serenova, LLC. All rights reserved.
 */

import React from 'react';
import CheckboxFilterMenu from '../../containers/CheckboxFilterMenu';

export default function(value, tableType) {
  return {
    Header: 'Skills',
    show: value,
    filterable: true,
    sortable: false,
    Filter: skillsFilter(tableType),
    id: 'skills',
    accessor: d => {
      let skillsArray = [];
      d.agents.forEach(agent =>
        agent.skills.forEach(skill => {
          if (skillsArray.indexOf(skill.skillName) === -1) {
            skillsArray.push(skill.skillName);
          }
        })
      );
      return skillsArray.join(', ');
    },
    Cell: ({ value }) => <span title={value}>{value}</span>
  };
}

export function skillsFilter(tableType) {
  return (
    <CheckboxFilterMenu
      menuType="Skills"
      tableType={tableType}
      buttonType="columnFilter"
      selectionType="checkbox"
    >
      Skills
    </CheckboxFilterMenu>
  );
}
