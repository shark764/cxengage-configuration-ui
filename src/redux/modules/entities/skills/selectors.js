/*
 * Copyright Â© 2015-2018 Serenova, LLC. All rights reserved.
 */
import { getCurrentForm } from '../../form/selectors';

export const getHasProficiencyFormValue = state =>
  getCurrentForm(state) && getCurrentForm(state).getIn(['values', 'hasProficiency']);
