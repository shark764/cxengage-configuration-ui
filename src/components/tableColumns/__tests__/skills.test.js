import skills from '../skills';

describe('skills table column', () => {
  it('function returns correct object', () => {
    expect(skills('Skills', 'mockTableType')).toMatchSnapshot();
  });
  it('accessor returns skills names as a string seperated by commas', () => {
    expect(
      skills('Skills', 'mockTableType').accessor({
        agents: [
          {
            skills: [{ skillName: 'mockSkill1' }, { skillName: 'mockSkill2' }]
          },
          {
            skills: [{ skillName: 'mockSkill3' }, { skillName: 'mockSkill4' }]
          },
          { skills: [{ skillName: 'mockSkill5' }, { skillName: 'mockSkill6' }] }
        ]
      })
    ).toMatchSnapshot();
  });
  it('accessor returns single skill name with no comma', () => {
    expect(
      skills('Skills', 'mockTableType').accessor({
        agents: [{ skills: [{ skillName: 'mockSkill1' }] }]
      })
    ).toMatchSnapshot();
  });
  it('accessor returns skills names as a string seperated by commas avoiding duplicate skills from multiple agents', () => {
    expect(
      skills('Skills', 'mockTableType').accessor({
        agents: [
          {
            skills: [{ skillName: 'mockSkill1' }, { skillName: 'mockSkill2' }]
          },
          {
            skills: [{ skillName: 'mockSkill1' }, { skillName: 'mockSkill2' }]
          },
          { skills: [{ skillName: 'mockSkill1' }, { skillName: 'mockSkill2' }] }
        ]
      })
    ).toMatchSnapshot();
  });
});
