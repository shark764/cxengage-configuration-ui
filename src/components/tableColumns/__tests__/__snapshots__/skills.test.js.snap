// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`skills table column Cell returns the value wrapped in a span 1`] = `
<span
  title="mockValue"
>
  mockValue
</span>
`;

exports[`skills table column accessor returns single skill name with no comma 1`] = `"mockSkill1"`;

exports[`skills table column accessor returns skills names as a string seperated by commas 1`] = `"mockSkill1, mockSkill2, mockSkill3, mockSkill4, mockSkill5, mockSkill6"`;

exports[`skills table column accessor returns skills names as a string seperated by commas avoiding duplicate skills from multiple agents 1`] = `"mockSkill1, mockSkill2"`;

exports[`skills table column function returns correct object 1`] = `
Object {
  "Cell": [Function],
  "Filter": <Connect(CheckboxMenu)
    buttonType="columnFilter"
    menuType="Skills"
    selectionType="checkbox"
    tableType="mockTableType"
  >
    Skills
  </Connect(CheckboxMenu)>,
  "Header": "Skills",
  "accessor": [Function],
  "filterable": true,
  "id": "skills",
  "show": "Skills",
  "sortable": false,
}
`;
